crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -49,6 +54,213 @@ pub trait WorldQuery {
+///
+/// The derive macro implements [`WorldQuery`] for your type and declares two structs that
+/// implement [`Fetch`] and [`FetchState`] and are used as [`WorldQuery::Fetch`] and
+/// [`WorldQuery::State`] associated types respectively.
+///
+/// **Note:** currently, the macro only supports named structs.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// use bevy_ecs::query::Fetch;
+///
+/// #[derive(Component)]
+/// struct Foo;
+/// #[derive(Component)]
+/// struct Bar;
+/// #[derive(Component)]
+/// struct OptionalFoo;
+/// #[derive(Component)]
+/// struct OptionalBar;
+///
+/// #[derive(Fetch)]
+/// struct MyQuery<'w> {
+///     entity: Entity,
+///     foo: &'w Foo,
+///     // `Mut<'w, T>` is a necessary replacement for `&'w mut T`
+///     bar: Mut<'w, Bar>,
+///     optional_foo: Option<&'w OptionalFoo>,
+///     optional_bar: Option<Mut<'w, OptionalBar>>,
+/// }
+///
+/// fn my_system(mut query: Query<MyQuery>) {
+///     for q in query.iter_mut() {
+///         q.foo;
+///     }
+/// }
+///
+/// # my_system.system();
+/// ```
+///
+/// ## Nested queries
+///
+/// Using nested queries enable the composition pattern, which makes it possible to re-use other
+/// query types. All types that implement [`WorldQuery`] (including the ones that use this derive
+/// macro) are supported.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// use bevy_ecs::query::Fetch;
+///
+/// #[derive(Component)]
+/// struct Foo;
+/// #[derive(Component)]
+/// struct Bar;
+/// #[derive(Component)]
+/// struct OptionalFoo;
+/// #[derive(Component)]
+/// struct OptionalBar;
+///
+/// #[derive(Fetch)]
+/// struct MyQuery<'w> {
+///     foo: FooQuery<'w>,
+///     bar: (&'w Bar, Option<&'w OptionalBar>)
+/// }
+///
+/// #[derive(Fetch)]
+/// struct FooQuery<'w> {
+///     foo: &'w Foo,
+///     optional_foo: Option<&'w OptionalFoo>,
+/// }
+///
+/// ```
+///
+/// ## Read-only queries
+///
+/// All queries that are derived with `Fetch` macro have their read-only variants with `ReadOnly`
+/// suffix. If you are going to use a query only for reading, you can mark it with `read_only`
+/// attribute.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// use bevy_ecs::query::{Fetch, ReadOnlyFetch, WorldQuery};
+///
+/// #[derive(Component)]
+/// struct Foo;
+/// #[derive(Component)]
+/// struct Bar;
+///
+/// #[derive(Fetch)]
+/// #[read_only]
+/// struct FooQuery<'w> {
+///     foo: &'w Foo,
+///     bar_query: BarQueryReadOnly<'w>,
+/// }
+///
+/// #[derive(Fetch)]
+/// struct BarQuery<'w> {
+///     bar: &'w Bar,
+/// }
+///
+/// fn assert_read_only<T: ReadOnlyFetch>() {}
+///
+/// assert_read_only::<<FooQuery as WorldQuery>::Fetch>();
+/// assert_read_only::<<BarQuery as WorldQuery>::ReadOnlyFetch>();
+/// // the following will fail to compile:
+/// // assert_read_only::<<BarQuery as WorldQuery>::Fetch>();
+/// ```
+///
+/// If you want to use derive macros with read-only query variants, you need to pass them with
```

Comment:
Hmm, this is a bit surprising. Can you explain why?

Reply:
As we generate a new struct with ReadOnly suffix, we need specify derives for it as well. Drive macros can't access the list of other drives specified by a user, this is why we need to use a separate attribute to pass them explicitly. I'll try to make this clearer in the docs.

Reply:
Updated the code snippet to make this clearer.